### PR TITLE
Kafka: improve support

### PR DESCRIPTION
**Pull request by pascalgrimaud**
Fix https://github.com/jhipster/generator-jhipster/issues/9286
Related to https://github.com/jhipster/generator-jhipster/issues/9287

I'm learning Kafka and following this blog https://www.confluent.io/blog/apache-kafka-spring-boot-application

If the docker-compose file is OK, I will update all configuration for Kubernetes, Openshift, etc.

It is a very first step.
Then, maybe we can push to specific topic for each entities ?

cc @jdubois @framiere @Dabz
_____ 

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by cbornet** [comment]
What would be great is to have something that shows up in the front end. For instance streaming the audit events either by websocket (if the option is used) or by http streaming.

**Comment by cbornet** [comment]
Or the tracking activity already managed by websockets.

**Comment by pascalgrimaud** [comment]
Yes it's a good idea @cbornet 
It would be cool to do something similar to the websocket page

**Comment by cbornet** [comment]
With the advantage that it would work with a multi-instance/scaled app.

**Comment by pascalgrimaud** [comment]
So for this first PR, let's focus only on backend. Improvements on Frontend can be done in another PR :D

I'll update the docker-compose files and Kubernetes config to use the new properties and this PR will be ready to be reviewed

**Comment by cbornet** [comment]
The problem with the generated producer, consumer and endpoints is that although they are a good example of working code, they are most probably code that people will not need and that they will need to remove. So I'm not sure that's a great dev experience...

**Comment by pascalgrimaud** [comment]
You're totally right, it should be better with entities but as I don't have any experience, I don't know how to handle it.

Here, it's a simple hello world of the use of producer / consumer. Nothing really exciting.

@cbornet : do you have experience with Kafka ?

**Comment by jdubois** [comment]
For me this is just a simple template with a working example, and that's fine for people to start customizing.

**Comment by cbornet** [comment]
@pascalgrimaud yes I have some experience with Kafka. We use it at work for a variety of use-cases. Actually, you can do a lot of different things with Kafka such as event bus, stream processing, event sourcing, ... That's why it's hard to provide a basic integration that would be interesting in all case.
I actually consider to migrate some of our Kafka Stream based micro-services to JHipster. This would add automatically some features that we require for such services : actuator endpoints (including prometheus), logger endpoint, dev/prod profiles, docker-compose setup, small admin UI, etc... But for that we also need to be able to generate apps without user management, without DB, etc... Hence the PRs I've pushed recently. I think JHipster would be valuable for purely stateless services/apps but we have some cleanup work to do first.
As for Kafka, the most interesting parts we can generate I think, is dependencies, Spring Boot autoconf, dev/prod profiles, and test conf (BTW there is probably some conf to do to configure an embedded kafka in test). The rest is not that useful because we don't know if users will do Kafka Streams, KSQL, KafkaConsumer/KafkaProducer or KafkaTemplate.

**Comment by pascalgrimaud** [comment]
There are already a lot of changes.
I think it is enough for a 1st improvement for Kafka.

What could be done later:
- a UI / page for producing and seeing new message
- something related to websocket ?
- maybe some code related to entities ?

File: generators/server/templates/src/main/java/package/web/rest/KafkaResource.java.ejs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package <%=packageName%>.web.rest;
+
+import <%=packageName%>.service.SampleKafkaProducer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping(value = "/api/kafka")
+public class KafkaResource {
```
Comment (cbornet) [inline generators/server/templates/src/main/java/package/web/rest/KafkaResource.java.ejs]: Maybe add an test for this class ?

File: generators/server/templates/src/test/java/package/CassandraKeyspaceIT.java.ejs [hunk 2]
```diff
@@ -23,9 +23,15 @@ import com.datastax.driver.core.Session;
...
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
+<%_ if (messageBroker === 'kafka') { _%>
+import org.springframework.kafka.test.context.EmbeddedKafka;
+<%_ } _%>
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+<%_ if (messageBroker === 'kafka') { _%>
+@EmbeddedKafka
```
Comment (cbornet) [inline generators/server/templates/src/test/java/package/CassandraKeyspaceIT.java.ejs]: Is it needed to have an EmbeddedKafka in all tests ?

Comment (pascalgrimaud) [reply generators/server/templates/src/test/java/package/CassandraKeyspaceIT.java.ejs]: Thanks a lot for starting to review my PR @cbornet 

I pushed quickly what I did before vacations, so not really sure about this part.
And I don't like this part, as I added `@EmbeddedKafka` everywhere. I started to use a dedicated `@Configuration` in `EmbeddedKafkaConfiguration` but I needed to customize the Embedded Kafka myself and didn't manage to make it works.

So ideas are welcome here

File: generators/server/templates/src/main/java/package/web/rest/KafkaResource.java.ejs (hunk 1 above)
Comment (pascalgrimaud) [reply generators/server/templates/src/main/java/package/web/rest/KafkaResource.java.ejs]: sure, will do

**Comment by pascalgrimaud** [comment]
Just pushed sample projects to play with:
https://github.com/pascalgrimaud/jhipster-kafka-samples

Tested with Docker-Compose and Kubernetes -> everything looks OK for me

**Comment by pascalgrimaud** [comment]
@jdubois : related to your tweet https://twitter.com/juliendubois/status/1126192008283004930?s=19
Just keep in mind, this PR will remove spring-cloud-stream dependencies

**Comment by vwiencek** [comment]
Hello, actually I see a use-case for Kafka with microservices / hipster. In a Microservice architecture, we often have :
- one main `User` entity holding credential information in a user management microservice
- for all other microservices, specific service related `User` entities (like `serviceAUser` or `serviceBUser` holding complementary information

Then each time we modify the main `User` entity (for example `email` or `firstname`, we need to keep other `serviceXUser` in sync with the main one.

The user management service might send an event to the kafka bus service that will be consumed exactly once by other microservices.

**Comment by mraible** [comment]
Good idea. Or you could use oauth2 for the authentication type and let your identity provider do this for you.

**Comment by pascalgrimaud** [comment]
Yes, there are a lot of use cases with Kafka, but I'd prefer to keep PR as little as possible.
This one is already big, so once it's ok and merged, we can improve some other use case.

**Comment by pascalgrimaud** [comment]
Unless someone is against this, I think I will merge this PR.
For me, it's far better than what we currently have. So then, we can do some little PR to advance steps by steps.

If someone is against, let me know !

**Pull request merged**